pkg/client/ssa_client.go

```diff
@@ -0,0 +1,218 @@
+func (c *ssaApplyObjectConfiguration) Configure(obj client.Object, s *runtime.Scheme) error {
+	if c.owner != nil {
+		if err := controllerutil.SetControllerReference(c.owner, obj, s); err != nil {
+			return err
+		}
+	}
+	MergeLabels(obj, c.newLabels)
+
+	return nil
+}
+
+// ApplyObject creates the object if is missing and if the owner object is provided, then it's set as a controller reference.
+// If the objects exists then when the spec content has changed (based on the content of the annotation in the original object) then it
+// is automatically updated. If it looks to be same then based on the value of forceUpdate param it updates the object or not.
+// The return boolean says if the object was either created or updated (`true`). If nothing changed (ie, the generation was not
+// incremented by the server), then it returns `false`.
+//
+// NOTE: the return boolean IS ALWAYS TRUE on success by default. This is much more efficient. If you truly need
+// to determine whether the object has been updated or not (which is not the case anywhere in the codebase but the tests),
+// you can use the DetermineUpdate() option.
+func (c *SsaApplyClient) ApplyObject(ctx context.Context, obj client.Object, options ...SsaApplyObjectOption) (bool, error) {
+	config := newSsaApplyObjectConfiguration(options...)
+	if err := config.Configure(obj, c.Client.Scheme()); err != nil {
+		return false, err
+	}
+
+	if err := prepareForSSA(obj, c.Client.Scheme()); err != nil {
+		return false, err
+	}
+
+	if config.skipIf != nil && config.skipIf(obj) {
+		return false, nil
+	}
+
+	updated := true
+
+	// NOTE: once the SSA migration is not needed anymore, read the orig conditionally only when config.determineUpdate is true.
+
+	// this cannot fail - we would have failed previously already because the scheme wouldn't have known the object's kind
+	origRo, _ := c.Client.Scheme().New(obj.GetObjectKind().GroupVersionKind())
+	orig := origRo.(client.Object)
+
+	if err := c.Client.Get(ctx, client.ObjectKeyFromObject(obj), orig); err != nil && !apierrors.IsNotFound(err) {
+		return false, err
+	}
+
+	// NOTE: remove this once this code is used for all "apply-style" functionality and has updated all the objects in the cluster.
+	if isSsaMigrationNeeded(orig, c.NonSSAFieldOwner) {
+		if err := c.MigrateToSSA(ctx, orig); err != nil {
+			return false, err
+		}
+	}
+
+	if err := c.Client.Patch(ctx, obj, client.Apply, client.FieldOwner(c.FieldOwner), client.ForceOwnership); err != nil {
+		return false, fmt.Errorf("unable to patch '%s' called '%s' in namespace '%s': %w", obj.GetObjectKind().GroupVersionKind(), obj.GetName(), obj.GetNamespace(), err)
+	}
+
+	if config.determineUpdate {
+		updated = obj.GetGeneration() != orig.GetGeneration()
+	}
+
+	return updated, nil
+}
+
+func prepareForSSA(obj client.Object, scheme *runtime.Scheme) error {
+	gvk, err := apiutil.GVKForObject(obj, scheme)
+	if err != nil {
+		return err
+	}
+	obj.GetObjectKind().SetGroupVersionKind(gvk)
```

Comment:
what if the GVK of the object is not in the scheme? We cannot expect that all GVKs of all resources defined in the templates would be present in the scheme (eg currently the appstudio API).
TBH, in most cases (when the object is taken from a template) the GVK should be already present in the object. Couldn't we simply check if it's already set or not, and try to set it from the scheme only if it's not set in the object already?

Reply:
If the GVK is not in the scheme, the call to client.Patch, Create, Get, ... will fail regardless as far as I am aware. I think we would have to work with a dynamic client to support that and then it would become problem of determining the GVR (group version resource) from the object itself, which is not possible in the general sense because the resource string is different from kind (usually, its just its plural, but we've seen with kubevirt and their sub-resources that stuff can get complex).

Reply:
Ok, I learned something today :) Fix incoming :)